packages/contracts/contracts/CyberMarketplace.sol

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import '@openzeppelin/contracts/utils/Counters.sol';
+import '@openzeppelin/contracts/utils/Address.sol';
+import '@openzeppelin/contracts/token/ERC1155/IERC1155.sol';
+import '@openzeppelin/contracts/utils/introspection/ERC165Checker.sol';
+import '@openzeppelin/contracts/token/ERC1155/utils/ERC1155Holder.sol';
+
+contract CyberMarketplace is ERC1155Holder {
+  using Counters for Counters.Counter;
+
+  struct Listing {
+    uint256 id;
+    address nftContract;
+    address payable seller;
+    uint256 tokenId;
+    uint256 price;
+    uint256 endTime;
+  }
+
+  event Sell(
+    uint256 indexed id,
+    address indexed nftContract,
+    uint256 indexed tokenId,
+    address seller,
+    uint256 price,
+    uint256 endTime
+  );
+
+  event Buy(
+    uint256 indexed id,
+    address indexed nftContract,
+    uint256 indexed tokenId,
+    address seller,
+    address buyer,
+    uint256 price
+  );
+
+  address[] private whitelistedContracts;
+
+  Counters.Counter private listingsId;
+
+  mapping(uint256 => Listing) public listings;
+
+  function isValidNFT(address nftContract) public view returns (bool) {
+    return ERC165Checker.supportsInterface(nftContract, type(IERC1155).interfaceId);
+  }
+
+  function init(address[] calldata _whitelist) public {
+    for(uint i = 0; i < _whitelist.length; i++) {
```

Comment:
use mapping to avoid increase of computation time, with the key => value mapping structure you avoid this loop.